Applications/Spire/Source/Charting/ChartView.cpp

```diff
@@ -546,19 +562,71 @@ void ChartView::update_auto_scale() {
     auto_scale_top = std::max(auto_scale_top, candle.GetHigh());
     auto_scale_bottom = std::min(auto_scale_bottom, candle.GetLow());
   }
-  set_region({m_top_left.m_x, auto_scale_top},
-    {m_bottom_right.m_x, auto_scale_bottom});
+  m_region.m_top_left.m_y = auto_scale_top;
+  m_region.m_bottom_right.m_y = auto_scale_bottom;
+  update_origins();
 }
 
-void ChartView::update_gaps() {
-  m_gaps.clear();
-  for(auto i = m_candlesticks.begin(); i < m_candlesticks.end() - 1; ++i) {
-    auto end = i->GetEnd();
-    auto next_start = std::next(i)->GetStart();
-    if(end != next_start) {
-      m_gaps.push_back({end, next_start});
+ChartView::GapInfo ChartView::update_gaps(std::vector<ChartView::Gap>& gaps,
+    std::vector<Candlestick>& candlesticks, ChartValue start) {
+  auto gap_info = GapInfo{0, ChartValue()};
+  for(auto& candlestick : candlesticks) {
+    auto end = candlestick.GetStart();
+    if(start != end) {
+      gaps.push_back({start, end});
+      gap_info.total_gaps_value += end - start;
+      ++gap_info.gap_count;
     }
+    start = candlestick.GetEnd();
   }
+  return gap_info;
+}
+
+QtPromise<ChartView::LoadedData> ChartView::load_data(
+    QtPromise<std::vector<Candlestick>>& promise, LoadedData data,
+    ChartModel* model) {
+  return promise.then([=] (auto result) mutable {
+    auto new_candlesticks = std::move(result.Get());
+    if(!new_candlesticks.empty() && !data.m_candlesticks.empty() &&
+        new_candlesticks.front().GetStart() <=
+        data.m_candlesticks.back().GetStart()) {
+      auto last = std::find_if(new_candlesticks.begin(),
+        new_candlesticks.end(), [&] (const auto& candlestick) {
+          return candlestick.GetStart() >=
+            data.m_candlesticks.back().GetStart();
+        });
+      auto b = new_candlesticks.begin();
+      if(last != new_candlesticks.end()) {
+        last++;
+      }
+      new_candlesticks.erase(b, last);
+    }
+    auto last = [&] {
+      if(data.m_candlesticks.empty() && new_candlesticks.empty()) {
+        return ChartValue();
+      } else if(data.m_candlesticks.empty()) {
+        return new_candlesticks.front().GetStart();
+      }
+      return data.m_candlesticks.back().GetEnd();
+    }();
+    auto gap_info = update_gaps(data.m_gaps, new_candlesticks, last);
+    data.m_candlesticks.insert(data.m_candlesticks.end(),
+      new_candlesticks.begin(), new_candlesticks.end());
+    data.m_current_x += std::min(static_cast<int>(std::ceil((
+      data.m_end - data.m_start - gap_info.total_gaps_value) /
+      data.m_values_per_pixel + gap_info.gap_count * GAP_SIZE())),
+      data.m_end_x);
+    data.m_start = data.m_end;
+    data.m_end += (data.m_end_x - data.m_current_x) *
+      data.m_values_per_pixel;
+    if(data.m_current_x < data.m_end_x) {
+      auto new_promise = model->load(data.m_start, data.m_end);
+      return load_data(new_promise, std::move(data), model);
+    }
+    return QtPromise<LoadedData>([=] {
```

Comment:
This is making an unneccesary copy of ```data```.

Reply:
What's the difference between:

Promise([data_copy = std::move(data); ] { return std::move(data_copy); });

and:

Pormise([&] { return std::move(data); });

The second one was my initial idea, but I came across the first one on SO.

Reply:
The first one is semantically correct because no copy of ```data``` is ever made. The second one is a crash waiting to happen because while there is no copy, you are capturing ```data``` by reference and that reference will get invalidated as soon as the function goes out of scope.